From dafa53dab616b2b082267aa022bf63cb7404968e Mon Sep 17 00:00:00 2001
From: Gunther Hagleitner <gunther@apache.org>
Date: Thu, 4 Jun 2015 16:14:25 -0700
Subject: [PATCH 0583/1164] CDH-39759 : Backport HIVE-10934: Restore support
 for DROP PARTITION PURGE (Eugene Koifman via
 Gunther Hagleitner)

Note that the q-file in this patch is slightly different from upstream.
HIVE-9481 (allow column list specification in INSERT statement_
is not in CDH. So we are working around that in CDH by using an alternative
INSERT syntax.

(cherry picked from commit bef52661cdacc22ea6dc06d7b09cad49a63d0bf9)

Change-Id: I16bfccebf8bfbdf8aa6d35435a281689320e0dca
---
 .../org/apache/hadoop/hive/ql/parse/HiveParser.g   |    4 +-
 .../clientpositive/alter_table_partition_drop.q    |   16 ++++
 .../alter_table_partition_drop.q.out               |   87 ++++++++++++++++++++
 3 files changed, 105 insertions(+), 2 deletions(-)
 create mode 100644 ql/src/test/queries/clientpositive/alter_table_partition_drop.q
 create mode 100644 ql/src/test/results/clientpositive/alter_table_partition_drop.q.out

diff --git a/ql/src/java/org/apache/hadoop/hive/ql/parse/HiveParser.g b/ql/src/java/org/apache/hadoop/hive/ql/parse/HiveParser.g
index b85c7d3..c692d92 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/parse/HiveParser.g
+++ b/ql/src/java/org/apache/hadoop/hive/ql/parse/HiveParser.g
@@ -1130,8 +1130,8 @@ partitionLocation
 alterStatementSuffixDropPartitions[boolean table]
 @init { pushMsg("drop partition statement", state); }
 @after { popMsg(state); }
-    : KW_DROP ifExists? dropPartitionSpec (COMMA dropPartitionSpec)* ignoreProtection? replicationClause?
-    -> { table }? ^(TOK_ALTERTABLE_DROPPARTS dropPartitionSpec+ ifExists? ignoreProtection? replicationClause?)
+    : KW_DROP ifExists? dropPartitionSpec (COMMA dropPartitionSpec)* ignoreProtection? KW_PURGE? replicationClause?
+    -> { table }? ^(TOK_ALTERTABLE_DROPPARTS dropPartitionSpec+ ifExists? ignoreProtection? KW_PURGE? replicationClause?)
     ->            ^(TOK_ALTERVIEW_DROPPARTS dropPartitionSpec+ ifExists? ignoreProtection? replicationClause?)
     ;
 
diff --git a/ql/src/test/queries/clientpositive/alter_table_partition_drop.q b/ql/src/test/queries/clientpositive/alter_table_partition_drop.q
new file mode 100644
index 0000000..2eca21c
--- /dev/null
+++ b/ql/src/test/queries/clientpositive/alter_table_partition_drop.q
@@ -0,0 +1,16 @@
+set hive.exec.dynamic.partition=true;  
+set hive.exec.dynamic.partition.mode=nonstrict;
+
+DROP TABLE IF EXISTS part_table PURGE;
+CREATE TABLE part_table (key INT, value STRING) partitioned by (p STRING);
+
+-- Note that HIVE-9481 allow column list specification in INSERT statement
+-- is not in CDH. So we are working around that in CDH.
+-- INSERT INTO part_table PARTITION(p)(p,key,value) values('2014-09-23', 1, 'foo'),('2014-09-24', 2, 'bar');
+INSERT INTO part_table PARTITION(p) select 1 as key, 'foo' as value, '2014-09-23' as p;
+INSERT INTO part_table PARTITION(p) select 2 as key, 'bar' as value, '2014-09-24' as p;
+SELECT * FROM part_table;
+ALTER TABLE part_table DROP PARTITION (p='2014-09-23');
+SELECT * FROM part_table;
+ALTER TABLE part_table DROP PARTITION (p='2014-09-24') PURGE;
+SELECT * FROM part_table;
diff --git a/ql/src/test/results/clientpositive/alter_table_partition_drop.q.out b/ql/src/test/results/clientpositive/alter_table_partition_drop.q.out
new file mode 100644
index 0000000..134f6a3
--- /dev/null
+++ b/ql/src/test/results/clientpositive/alter_table_partition_drop.q.out
@@ -0,0 +1,87 @@
+PREHOOK: query: DROP TABLE IF EXISTS part_table PURGE
+PREHOOK: type: DROPTABLE
+POSTHOOK: query: DROP TABLE IF EXISTS part_table PURGE
+POSTHOOK: type: DROPTABLE
+PREHOOK: query: CREATE TABLE part_table (key INT, value STRING) partitioned by (p STRING)
+PREHOOK: type: CREATETABLE
+PREHOOK: Output: database:default
+PREHOOK: Output: default@part_table
+POSTHOOK: query: CREATE TABLE part_table (key INT, value STRING) partitioned by (p STRING)
+POSTHOOK: type: CREATETABLE
+POSTHOOK: Output: database:default
+POSTHOOK: Output: default@part_table
+PREHOOK: query: -- Note that HIVE-9481 allow column list specification in INSERT statement
+-- is not in CDH. So we are working around that in CDH.
+-- INSERT INTO part_table PARTITION(p)(p,key,value) values('2014-09-23', 1, 'foo'),('2014-09-24', 2, 'bar');
+INSERT INTO part_table PARTITION(p) select 1 as key, 'foo' as value, '2014-09-23' as p
+PREHOOK: type: QUERY
+PREHOOK: Input: _dummy_database@_dummy_table
+PREHOOK: Output: default@part_table
+POSTHOOK: query: -- Note that HIVE-9481 allow column list specification in INSERT statement
+-- is not in CDH. So we are working around that in CDH.
+-- INSERT INTO part_table PARTITION(p)(p,key,value) values('2014-09-23', 1, 'foo'),('2014-09-24', 2, 'bar');
+INSERT INTO part_table PARTITION(p) select 1 as key, 'foo' as value, '2014-09-23' as p
+POSTHOOK: type: QUERY
+POSTHOOK: Input: _dummy_database@_dummy_table
+POSTHOOK: Output: default@part_table@p=2014-09-23
+POSTHOOK: Lineage: part_table PARTITION(p=2014-09-23).key SIMPLE []
+POSTHOOK: Lineage: part_table PARTITION(p=2014-09-23).value SIMPLE []
+PREHOOK: query: INSERT INTO part_table PARTITION(p) select 2 as key, 'bar' as value, '2014-09-24' as p
+PREHOOK: type: QUERY
+PREHOOK: Input: _dummy_database@_dummy_table
+PREHOOK: Output: default@part_table
+POSTHOOK: query: INSERT INTO part_table PARTITION(p) select 2 as key, 'bar' as value, '2014-09-24' as p
+POSTHOOK: type: QUERY
+POSTHOOK: Input: _dummy_database@_dummy_table
+POSTHOOK: Output: default@part_table@p=2014-09-24
+POSTHOOK: Lineage: part_table PARTITION(p=2014-09-24).key SIMPLE []
+POSTHOOK: Lineage: part_table PARTITION(p=2014-09-24).value SIMPLE []
+PREHOOK: query: SELECT * FROM part_table
+PREHOOK: type: QUERY
+PREHOOK: Input: default@part_table
+PREHOOK: Input: default@part_table@p=2014-09-23
+PREHOOK: Input: default@part_table@p=2014-09-24
+#### A masked pattern was here ####
+POSTHOOK: query: SELECT * FROM part_table
+POSTHOOK: type: QUERY
+POSTHOOK: Input: default@part_table
+POSTHOOK: Input: default@part_table@p=2014-09-23
+POSTHOOK: Input: default@part_table@p=2014-09-24
+#### A masked pattern was here ####
+1	foo	2014-09-23
+2	bar	2014-09-24
+PREHOOK: query: ALTER TABLE part_table DROP PARTITION (p='2014-09-23')
+PREHOOK: type: ALTERTABLE_DROPPARTS
+PREHOOK: Input: default@part_table
+PREHOOK: Output: default@part_table@p=2014-09-23
+POSTHOOK: query: ALTER TABLE part_table DROP PARTITION (p='2014-09-23')
+POSTHOOK: type: ALTERTABLE_DROPPARTS
+POSTHOOK: Input: default@part_table
+POSTHOOK: Output: default@part_table@p=2014-09-23
+PREHOOK: query: SELECT * FROM part_table
+PREHOOK: type: QUERY
+PREHOOK: Input: default@part_table
+PREHOOK: Input: default@part_table@p=2014-09-24
+#### A masked pattern was here ####
+POSTHOOK: query: SELECT * FROM part_table
+POSTHOOK: type: QUERY
+POSTHOOK: Input: default@part_table
+POSTHOOK: Input: default@part_table@p=2014-09-24
+#### A masked pattern was here ####
+2	bar	2014-09-24
+PREHOOK: query: ALTER TABLE part_table DROP PARTITION (p='2014-09-24') PURGE
+PREHOOK: type: ALTERTABLE_DROPPARTS
+PREHOOK: Input: default@part_table
+PREHOOK: Output: default@part_table@p=2014-09-24
+POSTHOOK: query: ALTER TABLE part_table DROP PARTITION (p='2014-09-24') PURGE
+POSTHOOK: type: ALTERTABLE_DROPPARTS
+POSTHOOK: Input: default@part_table
+POSTHOOK: Output: default@part_table@p=2014-09-24
+PREHOOK: query: SELECT * FROM part_table
+PREHOOK: type: QUERY
+PREHOOK: Input: default@part_table
+#### A masked pattern was here ####
+POSTHOOK: query: SELECT * FROM part_table
+POSTHOOK: type: QUERY
+POSTHOOK: Input: default@part_table
+#### A masked pattern was here ####
-- 
1.7.9.5

